### PR TITLE
New topic for creating ACL rule.

### PR DESCRIPTION
**Pull request by chandru-199**
## Purpose of this pull request

This pull request (PR) - New topic for creating ACL rule in magento.

## Affected DevDocs pages

-  https://devdocs.magento.com/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.html

whatsnew
Added new tutorial for creating [ACL rule](https://devdocs.magento.com/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.html).

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by chandru-199** [comment]
Thanks @rogyar, I pushed the requested changes.

**Comment by dobooth** [comment]
Thanks for the contribution, @chandran-wagento. Please add the link to the Affected Devdocs page section in the description.
@rogyar, when approving new topics or Major Updates, please ensure there is `whatsnew` information in the description.

**Comment by chandru-199** [comment]
Thanks @dobooth, I have made changes as you requested. Please check it again.

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+1. Navigate to **SYSTEM** > Permissions > **User Roles** 
+1. After clicking the Role Resources tab, select Resource Access as Custom.
+![Resource access as custom]({{ site.baseurl }}/common/images/ext-best-practices/resource-access-custom-create-acl-rule.png)
+1. Select the resources and save the role.
+![Resource tree]({{ site.baseurl }}/common/images/ext-best-practices/role-resources-create-acl-rule.png)
+
+##Step 3. Use ACL in Menu
+
+In your module create `menu.xml` file. We can hide the menu to the users by adding the `resource` attribute to `add` nodes in menu.xml.
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
In your module, create the `etc/adminhtml/menu.xml` file. This file defines a menu that will be hidden from unauthorized users. The `resource` attributes in the `add` nodes determine which resource each action accesses.
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+> `etc/adminhtml/menu.xml`
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: Delete this line.

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Backend:etc/menu.xsd">
+    <menu>
+     <add id="Vendor_MyModule::menu" title="Custom Menu" module="Vendor_MyModule" sortOrder="10" resource="Vendor_MyModule::menu"/>
+     <add id="Vendor_MyModule::create" title="Create" module="Vendor_MyModule" sortOrder="10" parent="Vendor_MyModule::menu" action="custommenu/create/index" resource="Vendor_MyModule::create"/>
+     <add id="Vendor_MyModule::delete" title="Delete" module="Vendor_MyModule" sortOrder="20" parent="Vendor_MyModule::menu" action="custommenu/delete/index" resource="Vendor_MyModule::delete"/>
+    </menu>
+</config>
+```
+
+##Step 4. Use ACL in Controllers
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
## Step 4. Restrict controller access
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+We can restrict the access to admin controllers by overriding the `_isAllowed` method of \Magento\Framework\App\Action\Action class.
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
We can restrict the access to admin controllers by overriding the `_isAllowed` method of the `\Magento\Framework\App\Action\Action` class.
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+> `Controller/Adminhtml/Create/Index.php`
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
Add the following to your module's `Controller/Adminhtml/Create/Index.php` file:
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+```php
+protected function _isAllowed()
+{
+ return $this->_authorization->isAllowed('Vendor_MyModule::create');
+}
+```
+
+> `Controller/Adminhtml/Delete/Index.php`
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
Add the following to your module's `Controller/Adminhtml/Delete/Index.php` file:
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+```php
+protected function _isAllowed()
+{
+ return $this->_authorization->isAllowed('Vendor_MyModule::delete');
+}
+```
+
+##Step 5. Use ACL in API endpoints
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
## Step 5. Restrict web API access
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+We can restrict the users to access API endpoints, using the ACL rule. When we creates Web API configuration file (webapi.xml), the rules defined in acl.xml can be use to restrict the access to API endpoints.
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]:
```suggestion
We can restrict users from accessing API endpoints by using the ACL rule. By creating a Web API configuration file (`etc/webapi.xml`), the rules defined in `acl.xml` can restrict the access to API endpoints.
```

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+> `etc/webapi.xml`
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: Delete this line

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    <route url="/V1/admin/create" method="POST">
+        <service class="Vendor\MyModule\Api\Create" method="execute"/>
+        <resources>
+            <resource ref="Vendor_MyModule::create" />
+        </resources>
+    </route>
+</routes>
+```
+
+If user doesn't have permission, the action page will be like below.
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: If the user does not have permissions, the action page displays an "Access denied" message.

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+![Not allowed user action page]({{ site.baseurl }}/common/images/ext-best-practices/not_allowed_users_create_acl_rule.png)
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: Delete this graphic.

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+<?xml version="1.0"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Backend:etc/menu.xsd">
+    <menu>
+     <add id="Vendor_MyModule::menu" title="Custom Menu" module="Vendor_MyModule" sortOrder="10" resource="Vendor_MyModule::menu"/>
+     <add id="Vendor_MyModule::create" title="Create" module="Vendor_MyModule" sortOrder="10" parent="Vendor_MyModule::menu" action="custommenu/create/index" resource="Vendor_MyModule::create"/>
+     <add id="Vendor_MyModule::delete" title="Delete" module="Vendor_MyModule" sortOrder="20" parent="Vendor_MyModule::menu" action="custommenu/delete/index" resource="Vendor_MyModule::delete"/>
+    </menu>
+</config>
+```
+
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: I think a graphic showing the menu would be helpful.

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
+---
+group: extension-best-practices
+title: Create Access Control List(ACL) Rule
+contributor_name: Ziffity
+contributor_link: https://www.Ziffity.com/
+---
+
+Access Control List(ACL) rules allow admin to limit the permissions of users in Magento. We can use ACL rules for authorize the users to access menu,controllers and API endpoints.
+
```
Comment (keharper) [inline src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: You need to explicitly state what you'll be doing in this example. You're creating three custom resources (Custom Menu, Create, Delete), creating a role that has access to these resources, and taking steps to restrict access by three entities (Admin users, controllers, and web APIs).

File: src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+                    <resource id="Vendor_MyModule::create" title="Create" sortOrder="0" />
+                    <resource id="Vendor_MyModule::delete" title="Delete" sortOrder="100" />
+                </resource>
+            </resource>
+        </resources>
+    </acl>
+</config>
+```
+
+##Step 2. Create user role with custom resources
```
[the remark this answers is not in the document]
Comment (keharper) [reply src/guides/v2.3/ext-best-practices/tutorials/create-access-control-list-rule.md]: Add a description of what you're doing in this step.

**Comment by chandru-199** [comment]
Thanks @dobooth, Changes have made as per your request. Please review it again.

**Comment by chandru-199** [comment]
Thanks @keharper. I have made changes as you requested.

**Comment by keharper** [comment]
@chandru-199 This PR is almost ready to be merged. Please add some links to related topics. Also, please optimize the screenshot files, if you haven't done so already.

**Comment by chandru-199** [comment]
Thanks @dobooth,I have added related topics. And I already optimized the image files. Please review it again.

**Comment by chandru-199** [comment]
Thanks @keharper . I have added related topics section. And I have already optimized all the images. Please review it again.

**Comment by keharper** [comment]
running tests

**Comment by keharper** [comment]
Thanks for contributing this new topic @chandru-199!

**Pull request merged**